mantidproject/mantid
# Conversion between image formats in tomography GUI

New tab in the tomography GUI. This is a utility to convert between image formats.
Some input test stacks come in different formats, and sometimes users need to convert stacks in FITS or other formats to a different format before loading them into third party software for reconstruction/visualization/segmentation/etc.

This depends on: 
- [x] #16020, which should be merged first.

**To test:**
- Check the code
- New unit tests should pass
- Check that the documentation reads well.
- For a functional test, grab these files: (ISIS network) `\\olympic\babylon5\Public\fedemp\IMAT\convert_images_dummy_test_example`
- Convert the files to your favorite output format (from the list that the interface supports). For example:
  - Open the custom interface: "Diffraction" => "Tomographic Reconstruction". Tab Convert. Select "Pick Format" as "FITS. Then point the input path to the "dummy_larmor" subdirectory of the test data. Select output "Format" as "png". You can leave the maximum search depth as 3 by default. Enter an output path. See that it generates output images for the input images, reconstructing the same tree of subdirectories in the outputs. The interface will pop up a message when the conversion is finished. You can also use input path "dummy_stfc", input format "png", output format "tiff".
  - Open/close the interface and see that the settings are remembered between sessions.
  - NXTomo format is disabled as input, as we don't have a LoadNXTomo and we may not add it unless there's interest in it.
  - The interface can become unresponsive if the amount of images is large. This is unfortunately as it stands now, but this covers current needs and I wouldn't fix this with a temporary solution but rather add the Load/SaveImage algorithms, see more comments below.
- If you're an aggressive tester, you can check that if you create a directory inside the input path and use it as output path, the interface will not fall in the trap and will  not produce outputs for the output path itself.
- Still, if the input and output path are the same, the output files will be generated "in-place" next to the input files.
- You could also check that if you select an input or output path, and remove one or both of them before clicking "convert", the interface will behave well.

Fixes #15942
Also fixes #11817

(The second ticket was a long standing one about documentation, and now that all the tabs of the interface are documented I'd close it). We'll need one to add a worked out example or similar, which cannot be done at the moment.

For normal image formats (PNG, TIFF, etc.) this uses Qt classes (QImageWriter, QImageReader), which has some limitations but is the temporary solution until we have algorithms `Load/SaveImage` (#6843)

Load/Saving of FITS files is well testes in unit and system tests. NXTomo saving is also tested. Saving/loading other formats is unfortunately not well tested, as this is being done via Qt classes at the moment. The next step before we can improve on that is to have Load/SaveImage algorithms.

Release notes: https://github.com/mantidproject/mantid/commit/253929fd99b20ee2475f1f3d004b456cda08f254

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
